allow to override the default os.Hostname() when creating PrometheusURLs

gesellix opened the pull request:
@juliusv PR for #716

juliusv commented on web/web.go:
```diff
@@ -240,9 +241,11 @@ func MustBuildServerURL(pathPrefix string) string {
-	hostname, err := os.Hostname()
-	if err != nil {
-		panic(err)
+	if hostname == "" {
```
The Travis build is failing with:

```
.build/root/gopath/src/github.com/prometheus/prometheus/web/web.go:244: invalid operation: hostname == "" (mismatched types *string and string)
.build/root/gopath/src/github.com/prometheus/prometheus/web/web.go:245: cannot assign string to hostname (type *string) in multiple assignment
```

This'll need to be:

```
if *hostname == "" {
	hostname, err := os.Hostname()
	// ...the same...
}
```

juliusv replied:
Please compile before submitting :)

gesellix replied:
oops... sorry, I had problems with my corporate proxy to fetch dependencies... "what could possibly go wrong".

brian-brazil commented on web/web.go:
```diff
@@ -240,9 +241,11 @@ func MustBuildServerURL(pathPrefix string) string {
+			panic(err)
+		}
 	}
 	return fmt.Sprintf("http://%s:%s%s", hostname, port, pathPrefix)
```
You might want to specify https too remove the port. I'd suggest adjust the flag name and behaviour accordingly

juliusv replied:
There's another `os.Hostname()` occurrence further up in the file which should actually also be replaced. It finds any scrape URLs that are actually pointing to the local host, but there we really need the hostname instead of a URL. So this might not be so workable as a single flag.

brian-brazil replied:
I'd suggest removing that behaviour actually, as that may ultimately cause confusion if I asked to scrape localhost but that's not what's showing up.

juliusv replied:
It will scrape localhost then, and it will also show that as localhost on the web UI, but the linkified version of the target URL will link to the configured hostname.

gesellix replied:
@brian-brazil @juliusv now I'm confused :) Should the "globalURL" behave identically like the "MustBuildServerURL" func? The `pathPrefix` would make a difference, so I suggest to extract a common `GetServerHostname` func which could be used from both existing functions?

juliusv replied:
I'd keep the `globalURL` function as it is, except for also using the `web.hostname` flag value instead of `os.Hostname()`. The `globalURL` function is used only for the actual clickable link of a target on the /status page of the Prometheus server. So that when Prometheus monitors itself or something else at localhost, the linkified version will still take you to the right place when looking at the status page remotely. We can't generally tell whether that global URL should also contain the path prefix or not - it would depend on whether the Prometheus is scraping something with a path prefix (like itself) or not... so I'd say just keep it as it is and just change how it gets its hostname.

gesellix commented:
I added a `getServerHostname` func to be used in both "MustBuildServerURL" and "globalURL". If preferred, I can squash the commits.

gesellix commented:
I guess, I need a lesson in setting up my local go environment. `go build` didn't fail, but travis-ci did.

gesellix commented:
sorry for the failing builds - I'll fix it at home in a working Go environment.

gesellix commented:
@juliusv looks like the simple new flag now doesn't break the build. I also updated the Dockerfile to reflect the recently changed config format and changed logging.

juliusv commented on Dockerfile:
```diff
@@ -20,7 +20,7 @@ RUN apk add --update -t build-deps go git mercurial vim \
     && cd tools/rule_checker && go build -o /bin/rule_checker && cd ../.. \
     && mkdir -p /etc/prometheus \
-    && mv ./documentation/examples/prometheus.conf /etc/prometheus/prometheus.conf \
+    && mv ./documentation/examples/prometheus.yml /etc/prometheus/prometheus.yml \
```
Thanks! These Dockerfile changes were already done in https://github.com/prometheus/prometheus/commit/6a2983c615f96f8c80c031e2405700cb2fc648cd in the meantime.

gesellix replied:
ah, I didn't rebase, thanks for the heads up!

juliusv commented on web/web.go:
```diff
@@ -176,7 +177,7 @@ func getTemplate(name string, pathPrefix string) (*template.Template, error) {
 		},
 		"globalURL": func(url string) string {
-			hostname, err := os.Hostname()
+			hostname, err := getServerHostname()
```
I'd just call it `hostname()` (using a `get` prefix is rather uncommon in Go-land in general)

gesellix replied:
`hostname` clashes with `hostname()`, doesn't it? I have to admit that my Go experience is quite short :-)
Making it uppercase would fix it, but as far as I know that would make the function public. Would you be ok with `Hostname()`?

juliusv commented on web/web.go:
```diff
@@ -240,9 +241,16 @@ func MustBuildServerURL(pathPrefix string) string {
+		return *hostname, nil
+	}
+	return os.Hostname()
+}
```
Please configure your editor to save text files with a newline character after the last line. This prevents all kinds of issues, like this warning here, noise in diffs, as well as http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline.

gesellix replied:
Yes, I'll do. Which editor do you use for Go projects? I tried Sublime Text and currently use IntelliJ (with the Go plugin).

fabxc replied:
If you are a bit more visual, Sublime is a decent choice. It has good Go plugins which also run `gofmt` whenever you save.

gesellix replied:
I'm not sure if I'm more visual (vi is probably too minimal for me), but I prefer to navigate through code using the keyboard, preferably with shortcuts. IntelliJ is the best tool I know so far regarding code navigation and code completion, but it feels too heavy for Go projects. For Sublime I didn't feel so comfortable, even with the `GoSublime` plugin. Do you use any additional plugins in Sublime?

juliusv replied:
I'm using vim with a ton of Go plugins and other plugins (Ctrl-P is awesome for fuzzy-finding for example: https://github.com/kien/ctrlp.vim, NERDTree for a file tree, etc.).

gesellix replied:
so I have to give it a try, thanks for your hints!

juliusv replied:
Take a look at https://github.com/fatih/vim-go - it seems to be the most popular one-stop suite of Go functionality for vim at the moment.

gesellix replied:
:+1:

juliusv commented:
:+1: otherwise

gesellix commented:
now the `hostname` flag with `os.Hostname()` fallback are available in `Hostname()`.

juliusv commented on web/web.go:
```diff
@@ -240,9 +241,16 @@ func MustBuildServerURL(pathPrefix string) string {
 	return fmt.Sprintf("http://%s:%s%s", hostname, port, pathPrefix)
 }
+
+func Hostname() (string, error) {
```
Ah right, I didn't see that you already had a variable named `hostname`. If it's defined and set to `hostname()` in the same line, it would actually work, but I agree it can be confusing. Capitalizing a function means exporting it from the package though, so we want to avoid that (only export what's needed). With that in mind, I guess a name like `buildHostname` (or even `getHostname` :P) would be ok.

gesellix replied:
*g*
I chose `getHostname` to keep the same pattern like `getTemplate` in the same package.

juliusv commented:
:+1: thanks!

gesellix commented:
thank you, too, especially for your patience discussing such a simple piece of code with a Go noob ;-)

juliusv commented:
@gesellix My pleasure :)
